### PR TITLE
Added infrastructure to support an Oryx Buildpack

### PR DESCRIPTION
**Pull request by dorfire**
1. Added a command: `oryx buildpack-detect ...`
2. Added two required images - one to run `pack`, another one that serves as a builder

### CODE REVIEW

**Comment by kichalla** [comment]
It would be easier to review this if we had a intro & design meeting regarding build-pack etc.

File: build/build-buildpacks-images.sh [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+declare -r REPO_DIR=$( cd $( dirname "$0" ) && cd .. && pwd )
+source $REPO_DIR/build/__variables.sh
+
+# Build base image for builder
+cd "$BUILD_IMAGES_BUILD_CONTEXT_DIR"
+docker build -f "$BUILDER_BASE_IMAGE_DOCKERFILE" -t $DOCKER_BUILDER_BASE_IMAGE_REPO:latest .
+
+cd /tmp
+
+$REPO_DIR/images/buildpack-builder/install-pack.sh
```
Comment (kichalla) [inline build/build-buildpacks-images.sh]: Based on this script content, this will only run on non-windows platforms?

File: images/buildpack-builder/builder.toml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+[stack]
```
Comment (kichalla) [inline images/buildpack-builder/builder.toml]: Are there any tests that can be written for the changes in this PR?

File: .gitignore [hunk 3]
```diff
@@ -18,7 +18,6 @@
 x64/
 x86/
 bld/
-[Bb]in/
```
[the remark this answers is not in the document]
Comment (dorfire) [reply .gitignore]: so that I can add the `bin/` directory under the buildpack.

File: build/build-buildpacks-images.sh (hunk 1 above)
Comment (dorfire) [reply build/build-buildpacks-images.sh]: Correct. In any case, `pack create-builder` is unsupported on Windows

File: images/buildpack-builder/builder-base.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,14 @@
+FROM oryxdevms/build:latest
+
+LABEL io.buildpacks.stack.id=com.microsoft.oryx.stack
+
+# Configure non-root user
+RUN groupadd -g 1002 oryx_group && \
```
[the remark this answers is not in the document]
Comment (dorfire) [reply images/buildpack-builder/builder-base.Dockerfile]: stack images should have a non-root user configured

File: build/build-buildpacks-images.sh [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+# Create builder
+builderName="$DOCKER_DEV_REPO_BASE/buildpack-builder"
+./pack create-builder $builderName \
+	--builder-config $REPO_DIR/images/buildpack-builder/builder.toml \
+	--no-pull
+
+# Remove pack everything that was added by it
```
Comment (gpcastro) [inline build/build-buildpacks-images.sh]: nit: pack _and_ everything...

File: .gitignore (hunk 3 above)
[the remark this answers is not in the document]
Comment (kichalla) [reply .gitignore]: But wouldn't this be a nuisance now that this folder would always show up as something that needs to be checked in. I should probably better understand the scenario.

File: src/BuildScriptGenerator/IBuildScriptGenerator.cs [hunk 6]
```diff
@@ -14,5 +17,12 @@ public interface IBuildScriptGenerator
         /// <param name="script">The generated script if the operation was successful.</param>
         /// <returns><c>true</c> if the operation was successful, <c>false</c> otherwise.</returns>
         bool TryGenerateBashScript(BuildScriptGeneratorContext scriptGeneratorContext, out string script);
+
+        /// <summary>
+        /// Determines which platforms can be used to build the given application.
+        /// </summary>
+        /// <param name="scriptGeneratorContext">The <see cref="BuildScriptGeneratorContext"/> with parameters for check.</param>
```
Comment (kichalla) [inline src/BuildScriptGenerator/IBuildScriptGenerator.cs]: nit: long line

File: src/BuildScriptGeneratorCli/Commands/BaseCommand.cs [hunk 7]
```diff
@@ -33,6 +33,7 @@ public int OnExecute(CommandLineApplication app, IConsole console)
             try
             {
                 _serviceProvider = GetServiceProvider();
+                _serviceProvider?.GetRequiredService<ILogger<BaseCommand>>()?.LogInformation("Oryx command line: {cmdLine}", Environment.CommandLine);
```
Comment (kichalla) [inline src/BuildScriptGeneratorCli/Commands/BaseCommand.cs]: nit: long line...can you take care of this for all the changes in your PR?

File: src/BuildScriptGenerator/IEnvironment.cs [hunk 8]
```diff
@@ -14,14 +14,13 @@ namespace Microsoft.Oryx.BuildScriptGenerator
...
         /// </summary>
-        /// <param name="name">Name of the environment variable. Is case-sensitive.</param>
+        /// <param name="name">Name of the environment variable. Case-sensitive.</param>
         /// <returns>
-        /// The value of the environment variable specified by variable,
-        /// or null if the environment variable is not found.
+        /// The value of the given environment variable, or null if the environment variable isn't found.
         /// </returns>
-        string GetEnvironmentVariable(string name);
+        string GetEnvironmentVariable(string name, string defaultValue = null);
```
Comment (kichalla) [inline src/BuildScriptGenerator/IEnvironment.cs]: Why this change? I do not see it being used in any of the code you have in this PR

File: images/buildpack-builder/builder.toml (hunk 2 above)
Comment (dorfire) [reply images/buildpack-builder/builder.toml]: Added @ `BuildpackDetectCommandTest`

File: images/buildpack-builder/install-pack.sh [hunk 9]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+
```
Comment (arroyc) [inline images/buildpack-builder/install-pack.sh]: nit: do we need Microsoft mit licensing info her like we do in other scripts?

File: src/BuildScriptGeneratorCli/Commands/BuildpackDetectCommand.cs [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+[assembly: InternalsVisibleTo("DynamicProxyGenAssembly2, PublicKey=0024000004800000940000000602000000240000525341310004000001000100c547cac37abd99c8db225ef2f6c8a3602f3b3606cc9891605d02baa56104f4cfc0734aa39b93bf7852f7d9266654753cc297e7d2edfe0bac1cdcf9f717241550e0a7b191195b7667bb4f64bcb8e2121380fd1d9d46ad2d92d2d15605093924cceaf74c4861eff62abf69b9291ed0a340e113be11e6a7d3113e92484cf7045cc7")]
+
+namespace Microsoft.Oryx.BuildScriptGeneratorCli
+{
+    [Command("buildpack-detect", Description = "Determines whether Oryx can be applied as a buildpack to " +
+        "an app in the current working directory.")]
+    internal class BuildpackDetectCommand : CommandBase
+    {
+        public const int DetectorFailCode = 100;
```
Comment (arroyc) [inline src/BuildScriptGeneratorCli/Commands/BuildpackDetectCommand.cs]: curious if there will be any other failcode with 100

File: tests/BuildScriptGeneratorCli.Tests/BuildpackDetectCommandTest.cs [hunk 11]
```diff
@@ -0,0 +1,67 @@
+﻿// --------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+// --------------------------------------------------------------------------------------------
+
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Oryx.BuildScriptGeneratorCli;
```
Comment (kichalla) [inline tests/BuildScriptGeneratorCli.Tests/BuildpackDetectCommandTest.cs]: nit: sort

File: src/BuildScriptGeneratorCli/Commands/BuildpackDetectCommand.cs (hunk 10 above)
Comment (dorfire) [reply src/BuildScriptGeneratorCli/Commands/BuildpackDetectCommand.cs]: what do you mean?

File: tests/BuildScriptGeneratorCli.Tests/BuildpackDetectCommandTest.cs [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+            var cmd = new BuildpackDetectCommand { PlatformDir = string.Empty, PlanPath = string.Empty };
+            var console = new TestConsole();
+            var svcProvider = GetServiceProvider(cmd);
+
+            // Act
+            int exitCode;
+            using (new CurrentDirectoryChange(_testDirPath))
+            {
+                // Assert with an empty app directory
+                Assert.Equal(BuildpackDetectCommand.DetectorFailCode, cmd.Execute(svcProvider, console));
```
Comment (kichalla) [inline tests/BuildScriptGeneratorCli.Tests/BuildpackDetectCommandTest.cs]: This should be a separate test IMO

File: src/Common/CurrentDirectoryChange.cs [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+using System;
+using System.IO;
+
+namespace Microsoft.Oryx.Common
+{
+    public class CurrentDirectoryChange : IDisposable
+    {
+        private readonly string _originalCurrentDirectory;
+
+        public CurrentDirectoryChange(string newCurrentDirectory)
```
Comment (gpcastro) [inline src/Common/CurrentDirectoryChange.cs]: I like this idea. What about removing the `Change` from the name, so we can call `using new CurrentDirectory("abc")`?

File: tests/BuildScriptGeneratorCli.Tests/BuildpackDetectCommandTest.cs [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+        public void Execute_OutputsNode_WhenPackageJsonExists()
+        {
+            // Arrange
+            var cmd = new BuildpackDetectCommand { PlatformDir = string.Empty, PlanPath = string.Empty };
+            var console = new TestConsole();
+            var svcProvider = GetServiceProvider(cmd);
+
+            // Act
+            int exitCode;
+            using (new CurrentDirectoryChange(_testDirPath))
```
Comment (kichalla) [inline tests/BuildScriptGeneratorCli.Tests/BuildpackDetectCommandTest.cs]: I am just wondering if this would have wider impact...like tests in different classes run in parallel and if this would cause  breaks which are subtle...is there any other way to test this? Should/is this command always supposed to run from current directory only?

File: src/BuildScriptGenerator/IEnvironment.cs (hunk 8 above)
Comment (dorfire) [reply src/BuildScriptGenerator/IEnvironment.cs]: I guess I used it somewhere and then changed the underlying code..
In any case, wouldn't hurt to have it

File: src/Common/CurrentDirectoryChange.cs (hunk 13 above)
Comment (dorfire) [reply src/Common/CurrentDirectoryChange.cs]: I decided to remove this, based on Kiran's comment, and pass the source path to Oryx as an argument - just like in the `build` command.

File: src/BuildScriptGeneratorCli/Commands/BuildpackDetectCommand.cs [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+
+        [Option("--plan-path <path>", CommandOptionType.SingleValue, Description = "Build plan TOML path.")]
+        public string PlanPath { get; set; }
+
+        internal override bool IsValidInput(IServiceProvider serviceProvider, IConsole console)
+        {
+            var result = true;
+            var logger = serviceProvider.GetService<ILogger<BuildpackDetectCommand>>();
+            
+            if (!Directory.Exists(SourceDir))
```
Comment (kichalla) [inline src/BuildScriptGeneratorCli/Commands/BuildpackDetectCommand.cs]: The properties which take paths can contain relative paths, `.` etc. so we need to sanitize them to be sure we do not do the wrong checks. You can get the `options` from the service provider here to get the sanitized input `var options = serviceProvider.GetRequiredService<IOptions<BuildScriptGeneratorOptions>>().Value;`...take a look at `BuildCommand` for reference

File: src/BuildScriptGeneratorCli/Commands/BuildpackDetectCommand.cs [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+                result = false;
+            }
+
+            return result;
+        }
+
+        internal override void ConfigureBuildScriptGeneratorOptions(BuildScriptGeneratorOptions options)
+        {
+            BuildScriptGeneratorOptionsHelper.ConfigureBuildScriptGeneratorOptions(
+                options, SourceDir, null, null, null, null, scriptOnly: false, null);
```
Comment (kichalla) [inline src/BuildScriptGeneratorCli/Commands/BuildpackDetectCommand.cs]: nit: used named params for `null` so that it is easy to know what that argument is for.

File: src/BuildScriptGeneratorCli/Commands/BuildpackDetectCommand.cs [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+using System.Linq;
+using System.Runtime.CompilerServices;
+using McMaster.Extensions.CommandLineUtils;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Oryx.BuildScriptGenerator;
+using Microsoft.Oryx.Common;
+
+[assembly: InternalsVisibleTo("DynamicProxyGenAssembly2, PublicKey=0024000004800000940000000602000000240000525341310004000001000100c547cac37abd99c8db225ef2f6c8a3602f3b3606cc9891605d02baa56104f4cfc0734aa39b93bf7852f7d9266654753cc297e7d2edfe0bac1cdcf9f717241550e0a7b191195b7667bb4f64bcb8e2121380fd1d9d46ad2d92d2d15605093924cceaf74c4861eff62abf69b9291ed0a340e113be11e6a7d3113e92484cf7045cc7")]
```
Comment (kichalla) [inline src/BuildScriptGeneratorCli/Commands/BuildpackDetectCommand.cs]: Curious, what is this for? Typically this kind of stuff goes into the `AssemblyInfo.cs` file.

File: tests/BuildScriptGeneratorCli.Tests/BuildpackDetectCommandTest.cs [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+            var svcProvider = GetServiceProvider(cmd);
+
+            // Act
+            int exitCode;
+
+            // Assert with an empty app directory
+            Assert.Equal(BuildpackDetectCommand.DetectorFailCode, cmd.Execute(svcProvider, console));
+
+            // Add file to app directory and assert again
+            File.WriteAllText(Path.Combine(_testDirPath, NodeConstants.PackageJsonFileName), "\n");
```
Comment (kichalla) [inline tests/BuildScriptGeneratorCli.Tests/BuildpackDetectCommandTest.cs]: Since `_testDirPath` is the root temporary directory for all the tests in this class, I would suggest to create a sub-dir under it to have this file. This way each test has its own sandboxed folder.

File: tests/BuildScriptGeneratorCli.Tests/BuildpackDetectCommandTest.cs [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+                PlanPath = string.Empty
+            };
+            var console = new TestConsole();
+            var svcProvider = GetServiceProvider(cmd);
+
+            // Act
+            int exitCode;
+
+            // Assert with an empty app directory
+            Assert.Equal(BuildpackDetectCommand.DetectorFailCode, cmd.Execute(svcProvider, console));
```
Comment (kichalla) [inline tests/BuildScriptGeneratorCli.Tests/BuildpackDetectCommandTest.cs]: Create a separate test case for this :-)

File: src/BuildScriptGeneratorCli/Commands/BuildCommand.cs [hunk 20]
```diff
@@ -62,6 +62,12 @@ internal class BuildCommand : BaseCommand
             Description = "The destination directory.")]
         public string DestinationDir { get; set; }
 
+        [Option(
+            "--bp-layers-dir <dir>",
+            CommandOptionType.SingleValue,
+            Description = "Path to the layers directory, if invoked from within a Buildpack.")]
+        public string BuildpackLayersDir { get; set; }
```
Comment (kichalla) [inline src/BuildScriptGeneratorCli/Commands/BuildCommand.cs]: I do not see changes related to this property in this class. I guess this is kind of Todo for future?

Comment (dorfire) [reply src/BuildScriptGeneratorCli/Commands/BuildCommand.cs]: this is actually a temp change that should have been reverted - thanks!

File: src/BuildScriptGeneratorCli/Commands/BuildpackDetectCommand.cs (hunk 17 above)
Comment (dorfire) [reply src/BuildScriptGeneratorCli/Commands/BuildpackDetectCommand.cs]: This is another leftover from when I needed to mock some method. I removed it. Thanks again for the catch!

File: src/BuildScriptGeneratorCli/Commands/BuildpackDetectCommand.cs [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+            var logger = serviceProvider.GetService<ILogger<BuildpackDetectCommand>>();
+
+            if (!Directory.Exists(options.SourceDir))
+            {
+                logger.LogError("Could not find the source directory {srcDir}", options.SourceDir);
+                console.Error.WriteLine($"Error: Could not find the source directory '{options.SourceDir}'.");
+                result = false;
+            }
+
+            if (!File.Exists(PlanPath))
```
Comment (kichalla) [inline src/BuildScriptGeneratorCli/Commands/BuildpackDetectCommand.cs]: Sanitize this and as well as `PlatformDir` too

File: src/BuildScriptGeneratorCli/Commands/BuildpackDetectCommand.cs (hunk 10 above)
Comment (arroyc) [reply src/BuildScriptGeneratorCli/Commands/BuildpackDetectCommand.cs]: makes sense now .. thanks for adding the comment

File: src/BuildScriptGeneratorCli/Commands/BuildpackDetectCommand.cs (hunk 21 above)
Comment (dorfire) [reply src/BuildScriptGeneratorCli/Commands/BuildpackDetectCommand.cs]: you mean with `Path.GetFullPath`?

Comment (kichalla) [reply src/BuildScriptGeneratorCli/Commands/BuildpackDetectCommand.cs]: yes
